File: cuss/token.py
from enum import Enum, auto
from typing import Optional
import location

class TokenType(Enum):
    # simple textual tokens
    LPAREN = "("
    RPAREN = ")"
    LBRACE = "{"
    RBRACE = "}"
    LBRACKET = "["
    RBRACKET = "]"
    COMMA = ","
    DOT = "."
    COLON = ":"
    # arithmetic operators
    PLUS = "+"
    MINUS = "-"
    STAR = "*"
    SLASH = "/"
    MOD = "%"
    PLUS_EQ = "+="
    MINUS_EQ = "-="
    STAR_EQ = "*="
    SLASH_EQ = "/="
    MOD_EQ = "%="
    # comparison operators
    EQ_EQ = "=="
    NEQ = "!="
    LT = "<"
    LTE = "<="
    GT = ">"
    GTE = ">="
    # logical operators
    LOGICAL_AND = "&&"
    LOGICAL_OR = "||"
    LOGICAL_NOT = "!"
    # bitwise operators
    BITWISE_AND = "&"
    BITWISE_OR = "|"
    # assignment operators
    EQ = "="
    # keywords
    IF = "if"
    ELSE = "else"
    LOOP = "loop"
    WHILE = "while"
    FOR = "for"
    IN = "in"
    RETURN = "return"
    BREAK = "break"
    CONTINUE = "continue"
    TRUE = "true"
    FALSE = "false"
    FN = "fn"
    CONST = "const"
    VAR = "var"
    TYPE = "type"
    STRUCT = "struct"
    ENUM = "enum"
    # literals
    INTEGER = "integer"
    FLOAT = "float"
    STRING = "string"
    CHAR = "char"
    IDENT = "ident"
    # special tokens
    EOF = "eof"
    ERROR = "error"
    INDENT = "indent"
    DEDENT = "dedent"
    NEWLINE = "newline"
    

class Token:
    def __init__(self, type: TokenType, loc: location.Location, lexeme: Optional[str] = None) -> None:
        self.type = type
        self.lexeme = lexeme
        self.loc = loc
        
        
class SimpleToken(Token):
    def __init__(self, type: TokenType, lexeme: str, location: location.Location) -> None:
        super().__init__(type, lexeme, location)
        

File: src/frontend/token.cpp
#include "token.hpp"

const char* token_kind_to_string(TokenKind kind) {
    switch (kind) {
        case TokenKind::Identifier: return "Identifier";
        case TokenKind::Number: return "Number";
        case TokenKind::String: return "String";
        case TokenKind::Let: return "Let";
        case TokenKind::Fn: return "Fn";
        case TokenKind::If: return "If";
        case TokenKind::Else: return "Else";
        case TokenKind::While: return "While";
        case TokenKind::Loop: return "Loop";
        case TokenKind::Return: return "Return";
        case TokenKind::True: return "True";
        case TokenKind::False: return "False";
        case TokenKind::LParen: return "(";
        case TokenKind::RParen: return ")";
        case TokenKind::LBrace: return "{";
        case TokenKind::RBrace: return "}";
        case TokenKind::LBracket: return "[";
        case TokenKind::RBracket: return "]";
        case TokenKind::Comma: return ",";
        case TokenKind::Dot: return ".";
        case TokenKind::Colon: return ":";
        case TokenKind::Semicolon: return ";";
        case TokenKind::Plus: return "+";
        case TokenKind::Minus: return "-";
        case TokenKind::Star: return "*";
        case TokenKind::Slash: return "/";
        case TokenKind::Percent: return "%";
        case TokenKind::Eq: return "=";
        case TokenKind::EqEq: return "==";
        case TokenKind::Bang: return "!";
        case TokenKind::BangEq: return "!=";
        case TokenKind::Less: return "<";
        case TokenKind::LessEq: return "<=";
        case TokenKind::Greater: return ">";
        case TokenKind::GreaterEq: return ">=";
        case TokenKind::AmpAmp: return "&&";
        case TokenKind::BarBar: return "||";
        case TokenKind::Error: return "Error";
        case TokenKind::Eof: return "Eof";
        default: return "Unknown";
    }
}